### DISCUSSION TITLE
What is the plan for supporting build tools backends

### DISCUSSION THREAD

**Discussion by julie777** [pyscaffold/pyscaffold#618] [Dev Discussions]
I understood the comment about only supporting Git. Probably more than 80% of developers are using it.

We have multiple issues and discussions that reference `pyproject.toml`,  I spent the last couple days immersed in PyPA docs and all the PEPs that relate to `pyproject.toml`.

My conclusion is that as some point pyscaffold will need to support other build tools besides setuptools. Has there been any thinking or discussion about how that could be accomplished?

**Comment by abravalheri** [comment]
> pyscaffold will need to support other build tools besides setuptools

I think of that more of a "nice to have" than a "will need".
I was thinking that this could be accomplished via extensions (either distributed directly with PyScaffold, or 3rd Party).

E.g. probably we want to ship something like `putup --flit` in the future...

**Comment by FlorianWilhelm** [comment]
When `setuptools` switches to `pyproject.toml` what is the advantage of using any of the other backends like `flit`, `poetry`, etc. compared to it from a developers perspective?

**Reply by abravalheri** [reply]
Please find bellow my opinion about the comparison (from a dev's perspective):

- `flit`:
  1. Currently has better defaults than setuptools (e.g. `include_package_data`, automatic `packages` and `py_modules` discovery, no `MANIFEST.in`) and is easier to configure. Once setuptools supports PEP 621 and implements better defaults, the lines start to be a bit blurred (and I would argue that `setuptools` would be better because it also supports binary extensions and things like Cython).
  2. `flit` supports uploading packages, creating a very basic package skeleton, installing and building the package. It is not as *all-in-one* as `poetry`/`pdm`/`hatch` but its intention is to a more comprehensive tool (in opposition to a simple build backend).
    However my opinion is that this is less relevant if the dev is using something like PyScaffold, with a similar tox/nox setup.
  3. People say `flit` is faster... I say this does not mater a lot in the context of packaging.
  4. If `flit` has bugs, it is definitely easier to get started in `flit`'s code base and submit PR's. The maintainer's response is also faster, since there is more people interested in being maintainers (it is a virtuous cycle).

- `poetry`/`hatch`/`pdm`/...:
  1. These tools are much more than a packaging backend and try to be a `all-in-one` answer for Python's packaging. `poetry` (if not the others) also support building applications that are never meant to be packaged (e.g. they just need something with a lock file for dependencies).
  2. They are also easier to be used from people coming from other communities, e.g. Javascript, Ruby, ..., since they are more familiar with a `all-in-one` approach. Some newcomers might feel overwhelmed by the amount of tools used in traditional Python development.
  I also think this is less relevant if people are using PyScaffold, however PyScaffold will not help if the person is interested in just developing an app that does not need to be packaged.
  3. Each one of these solutions bring innovations and differentiating points, e.g. `pdm` will support [PEP 582](https://www.python.org/dev/peps/pep-0582/), which is very similar the way `.node_modules` work in JavaScript; `hatch` will do virtual env management; `poetry` has a custom dependency resolver, lock-file, etc...

**Reply by abravalheri** [reply]
- [`mesonpy`](https://meson-python.readthedocs.io/en/latest/) and other solutions:
  1. They are niche-specific and will solve specific problems (e.g. using `meson` for custom build steps/binary extensions).
  2. While `setuptools` does support binary extensions and custom build steps, it is not straight forward to do it, specially in complex code bases.
  3. Upstream compiles libraries can also depend on specific tools (e.g. `meson`, `cmake`, `conan`, etc...) so creating bindings might be just easier if you use such focused solutions.

**Reply by FlorianWilhelm** [reply]
Thanks for this extensive comparison list. 

We should also check if the tools support plugins. Setuptools-scm is very configurable by plugins and setuptools-scm makes use of this. As far as I know, the maintainer of `flit` is very much against this and thus developing something like `setuptool-scm` is not possible (without monkey-patching and this is a hack).

I have to confess that I secretly also played around with `poetry` a bit. I just used the building backend of it and checked if PyScaffold works with it. It works quite well actually by just changing in `pyproject.toml`:
```
[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```
and adding the `[tool.poetry...]` sections.

In contrast to `flit`, `poetry` is very much extensible with the help of plugins. There is also pendant for `setuptools-scm`, named [poetry-dynamic-versioning](https://pypi.org/project/poetry-dynamic-versioning/). With respect to building binary extensions, there is work undergoing in poetry for [build.py](https://github.com/python-poetry/poetry/issues/2740).

To sum these thoughts up, if we really want to replace `setuptools` with another build backend then I guess `poetry` is the most promising one.  Should we support several build backends at the same time? I think that would drastically increase the complexity. The build backend is the most fundamental part when building a package. There is also a good reason Linus decided to only have one scheduler in the kernel although several algorithms exist with different (dis-)advantages. He argued that it would just be too complex to support several at the same time.

**Reply by abravalheri** [reply]
PDM is actually quite good (and `hatch` seems promising, although I haven't played with it).

The main idea of PEP 621 is to allow people swapping between all these backends and playing with them. For example someone could start with `flit` for very simple packages but then swap for `setuptools` when they need to use Cython... `poetry` however seems to not support PEP 621 (maybe I am wrong here)[^1].

I would suggest that going forward we focus in developing solutions for backends that support somehow PEP 518, 621 and 660, i.e. the minimum denominator that guarantees compatibility/interchangeability. The challenge here is to find a replacement for `setuptools-scm` for each of them (`setuptools-scm` is such a cornerstone of PyScaffold!).

If someone wants to it should still possible to use PyScaffold's basic infrastructure/architecture (e.g. action pipeline, extensions, etc) to develop variations (e.g. for `poetry`), but I would argue that has to leave in a separated (very intrusive) extension that is not meant to interoperate with the other extensions.

[^1]: I really don't judge poetry's maintainers here, in the last few years there have been a lot of new PEPs "revolutionarising" the way packaging works in Python, and it is very overwhelming to keep up with them.

**Reply by FlorianWilhelm** [reply]
I think Cython is possible with poetry. You just do in `pyproject.toml`:
```
[tool.poetry]
build = "build.py"
```
and within `build.py` you call whatever needs to be done to compile your extensions, just like in `setup.py` before.

**Reply by abravalheri** [reply]
Sorry, I meant that `poetry` seems not to support PEP 621 (the original phrase was confusing, I corrected it now)

**Reply by FlorianWilhelm** [reply]
Ah, okay so what poetry has under `[tool.poetry]` would need to go into `[metadata]` or actually `[project]` (according to PEP 621)?

**Reply by abravalheri** [reply]
Just to wrap up my line of thought:

- When setuptools supports PEP 621, I think we can assume that swapping very "barebones" backends will work out of the box and even consider adding something like `--flit`.
- We can consider creating `pyscaffoldext-poetry`, but that will basically just reuse the very basic build blocks of PyScaffold and should not be considered to interoperate with the existing extensions.

Question, does `poetry` interoperates well with `conda`?

**Reply by abravalheri** [reply]
> Ah, okay so what poetry has under `[tool.poetry]` would need to go into `[metadata]` or actually `[project]` (according to PEP 621)?

Yes, if `poetry` was to support PEP 621, most of what is inside `[tool.poetry]` should go to `[project]` (after the required changes, because each table support different fields)

**Reply by FlorianWilhelm** [reply]
> Question, does poetry interoperates well with conda?

This depends. If we only use the build-backend of poetry then users can decide if they use poetry, pipenv or conda to manage their environments. Mixing things will not work I guess.
